examples/dtls-sock/dtls-client.c

```diff
@@ -151,39 +186,29 @@ static int client_send(char *addr_str, char *data, size_t datalen)
         return -1;
     }
 
-    res = credman_add(&credential0);
-    if (res < 0 && res != CREDMAN_EXIST) {
-        /* ignore duplicate credentials */
-        printf("Error cannot add credential to system: %d\n", (int)res);
-        return -1;
-    }
-
+    /*
+     * Currently DTLS sock needs one and only one credential for the
+     * initialization. Subsequent credentials are made available to the sock
+     * by means of `sock_dtls_add_credential`.
+     */
     if (sock_dtls_create(&dtls_sock, &udp_sock, SOCK_DTLS_CLIENT_TAG_0,
                          SOCK_DTLS_1_2, SOCK_DTLS_CLIENT) < 0) {
         puts("Error creating DTLS sock");
         sock_udp_close(&udp_sock);
         return -1;
     }
 
-#if IS_ACTIVE(CONFIG_DTLS_PSK)
-    /* register a second PSK credential */
-    res = credman_add(&credential1);
-    if (res < 0 && res != CREDMAN_EXIST) {
-        /* ignore duplicate credentials */
-        printf("Error cannot add credential to system: %d\n", (int)res);
-        return -1;
-    }
+    if (IS_ACTIVE(CONFIG_DTLS_PSK)) {
+        /* make the new credential available to the sock */
+        if (sock_dtls_add_credential(&dtls_sock, SOCK_DTLS_CLIENT_TAG_1) < 0) {
```

Comment:
> I was always very confused about what's the idea behind credman.

Yeah, I know what you mean...

> Shouldn't a credential be associated with a host? So what does it mean to add multiple credentials to a socket?

Not necessarily. I mean, you can already see the example of the server with more than one pair of keys. IMO one of the greatest problems of this module is that a credential is a single "private key" and 0 or more "client keys". Like I commented in the example, the way it currently works looks more like a keyring of client keys and your own.

> So what's the advantage of the credential IDs ('tags') as opposed to just passing pointers?

In key management using tags is not uncommon, this is crucial for instance when integrating secure elements or other backends where the key is directly not reachable, but a tag is passed to the crypto backend. I'm not sure why this was the case for credman, as it clearly doesn't support any other backend than a buffer in memory.

IMO we are in need of a rethinking of the credentials management in RIOT. We would really benefit from:
- not only DTLS credentials (of course)
- possibility of multiple backends including secure elements
- more types of credentials
- ability to indicate scope and capabilities of the credentials

Reply:
Hey @Einhornhool! Take a look at this comment please :tiger2: 
(Also, maybe this comment should be an issue instead?)

Reply:
Hmm most of the points in Leandros credentials requirement list could be handled by the PSA Crypto key management. It might be worth looking into patching tinydtls to use PSA for crypto operations and credential storage, instead of implementing the same features twice.